text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,214 @@
+  * All APIs that allow you immediate traversal up and down (e.g.: `elm.parentNode`) must be patched.
+  * All Event retargeting mechanism must be unwind to flatten them up.
+  * All deep traversal (e.g.: `elm.querySelector`) must be patched, and DOM filtering must be implemented by simulating the flatten representation of the application. This will require a query selector mechanism that can translate regular selectors into cross shadow selections (which is not trivial work)
+
+* Package the above functionality in a web component which semantically matches the `<script>` tag API.
+
+* Provide an additional mechanism to remap globals when needed.
+    * The sandbox does NOT include any global name defined in the main window, whether it is defined before or after the creation of the sandbox. But the `<virtual-script>` can opt-in to share a global value between the sandbox and the main window, this must be done via the `extra-globals` argument passed to the `<virtual-script extra-globals="dataLayer">` element.
+
+Other considerations for Phase II:
+
+ * the virtual script web component "should" be registered before any shadow root is created, it is not longer a "must", the flip side of this is that pre-existing shadowRoots are just not going to be inspected by virtual scripts.
+ * budget is less than 10kb (minify/gzip).
+ * performance penalties will be added to the virtual scripts as they run inside a membrane:
+   * evaluation budget: 30% slower than evaluating the same script in the main window.
+   * operation budget: 1X slower than executing the same task is the main window.
+   * 3rd party scripts are rarely part of the critical path for the application to function.
+ * synthetic shadow and native shadow combinations can be allowed, and supported, although it will require an explicit order to run before synthetic shadow polyfill does it.
+
+### Globals
+
+Some scripts define globals which are meant to be used by the rest of the application. Let's take this example of Google Analytics:
+
+```html
+<!-- Google Analytics -->
+<script>
+    window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</script>
+<script async src='https://www.google-analytics.com/analytics.js'></script>
+<!-- End Google Analytics -->
+```
+
+If we evaluate the above snippets with `<virtual-script>` instead, then `ga` global value will only be available into the sandbox. So `ga` will not be usable by anyone else in the entire application. We can however, allow for globals to be exposed into the main window. We call this global _exporting_. 
+
+Conversely, if a script running in `virtual-script` needs access to a global variable from the main window, we can _import_ this global. 
+
+A syntax like the following will be provided:
+
+```html
+<virtual-script exported-globals="foo"></virtual-script>
```

Comment:
They are read-only by design. @caridy can confirm if they are going to stay that way.

Reply:
How does the ability to modify the `globalThis` through `export-globals` align with Locker protections? Can `<virtual-script export-globals="...">` be used to pollute/poison the `globalThis` which is used by modules in multiple packages/namespaces on Salesforce platform?

Reply:
Yes we have talked about having an allow list of globals which can be mutated. I believe `globalThis` is one of those we need to protect against.

Reply:
That being said, we are not really setting out to provide tons of protection against malicious scripts, since only application owners can use `virtual-script` and the same security concerns exist with using a regular script tag.

Reply:
@tedconn remember that we now have two equivalent/reflective names: `exported-global-names` and `imported-global-names`.

@matt-buland-sfdc, they are analogous to import and exports in modules, the imported is readOnly inside the sandbox, and the exported is readOnly by the main app. 

@jbenallen in principle, the script evaluated inside virtual script is a high-privilege script, no restrictions on what it can do, therefore, it can attempt to do, and can do damage on the main window (e.g.: full access to cookies, etc). This is not a secure sandbox. neither imported and exported names are live values, they are defined by the person who insert the virtual script in the main window, and cannot be modified causing expansion of their initial value from within the sandbox itself, so, in that regards, it is all fine.